nbgrader/preprocessors/instantiatetests.py

```diff
@@ -0,0 +1,420 @@
+                new_lines.append('')
+
+        # run the trailing non-autotest lines, if any remain
+        if len(non_autotest_code_lines) > 0:
+            self._execute_code_snippet("\n".join(non_autotest_code_lines))
+
+        # add the final success message
+        if is_grade_flag and self.global_tests_loaded:
+            if self.autotest_delimiter in cell.source:
+                new_lines.append(self.success_code)
+
+        # replace the cell source
+        cell.source = "\n".join(new_lines)
+
+        # remove the execution metainfo
+        cell.pop('execution', None)
+
+        return cell, resources
+
+    # -------------------------------------------------------------------------------------
+    def _load_test_template_file(self, resources):
+        """
+        attempts to load the tests.yml file within the assignment directory. In case such file is not found
+        or perhaps cannot be loaded, it will attempt to load the default_tests.yaml file with the course_directory
+        """
+        self.log.debug('loading template tests.yml...')
+        self.log.debug(f'kernel_name: {resources["kernel_name"]}')
+        try:
+            with open(os.path.join(resources['metadata']['path'], self.autotest_filename), 'r') as tests_file:
+                tests = yaml.safe_load(tests_file)
+            self.log.debug(tests)
+
+        except FileNotFoundError:
+            # if there is no tests file, just load a default tests dict
+            self.log.warning(
+                'No tests.yml file found in the assignment directory. Loading the default tests.yml file in the course root directory')
+            # tests = {}
+            try:
+                with open(os.path.join(self.autotest_filename), 'r') as tests_file:
+                    tests = yaml.safe_load(tests_file)
+            except FileNotFoundError:
+                # if there is no tests file, just create a default empty tests dict
+                self.log.warning(
+                    'No tests.yml file found. If AUTOTESTS appears in testing cells, an error will be thrown.')
+                tests = {}
+            except yaml.parser.ParserError as e:
+                self.log.error('tests.yml contains invalid YAML code.')
+                self.log.error(e.msg)
+                raise
+
+        except yaml.parser.ParserError as e:
+            self.log.error('tests.yml contains invalid YAML code.')
+            self.log.error(e.msg)
+            raise
+
+        # get kernel specific data
+        tests = tests[resources["kernel_name"]]
+
+        # get the test templates
+        self.test_templates_by_type = tests['templates']
+
+        # get the test dispatch code template
+        self.dispatch_template = tests['dispatch']
+
+        # get the success message template
+        self.success_code = tests['success']
+
+        # get the hash code template
+        self.hash_template = tests['hash']
```

Comment:
I think it should be optional, and should generate an error if `HASHED` is required and this not filled.